pkg/scrape/pprof.go

```diff
@@ -15,267 +15,93 @@
 package scrape
 
 import (
-	"encoding/binary"
-	"strings"
 	"time"
 
-	"github.com/cespare/xxhash/v2"
-	"github.com/valyala/bytebufferpool"
-	"google.golang.org/protobuf/proto"
-
-	"github.com/pyroscope-io/pyroscope/pkg/agent/upstream"
 	"github.com/pyroscope-io/pyroscope/pkg/scrape/config"
 	"github.com/pyroscope-io/pyroscope/pkg/scrape/labels"
 	"github.com/pyroscope-io/pyroscope/pkg/scrape/model"
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/segment"
 	"github.com/pyroscope-io/pyroscope/pkg/storage/tree"
-	"github.com/pyroscope-io/pyroscope/pkg/structs/sortedmap"
-	"github.com/pyroscope-io/pyroscope/pkg/structs/transporttrie"
 )
 
 type pprofWriter struct {
-	// Instead of the upstream (which requires transporttrie.Trie)
-	// we could have a more generic samples consumer.
-	upstream upstream.Upstream
+	ingester Ingester
 	labels   labels.Labels
 	config   *config.Profile
-	time     time.Time
-	cache    cache
+
+	r *tree.ProfileReader
 }
 
-func newPprofWriter(u upstream.Upstream, target *Target) *pprofWriter {
-	return &pprofWriter{
-		upstream: u,
+func newPprofWriter(ingester Ingester, target *Target) *pprofWriter {
+	w := pprofWriter{
+		ingester: ingester,
 		labels:   target.Labels(),
 		config:   target.profile,
-		cache:    make(cache),
 	}
+	w.r = tree.NewProfileReader().SampleTypeFilter(w.filterSampleType)
+	return &w
 }
 
-func (w *pprofWriter) reset() {
-	w.time = time.Time{}
-	w.cache = make(cache)
-}
-
-func (w *pprofWriter) writeProfile(b []byte) error {
-	var p tree.Profile
-	if err := proto.Unmarshal(b, &p); err != nil {
+func (w *pprofWriter) writeProfile(st, et time.Time, b []byte) error {
+	p := tree.ProfileFromVTPool()
+	defer p.ReturnToVTPool()
+	p.Reset()
+	if err := p.UnmarshalVT(b); err != nil {
```

Comment:
I'm curious about performance improvement here (in case you have measured it).

Reply:
Working on it right now :)

Reply:
I've added some measurements - not final ones, just to give the idea.

Reply:
Awesome improvements! :clap::clap: